types/node/test.d.ts

```diff
@@ -1050,3 +966,183 @@ declare module 'node:test' {
+    /**
+     * The nesting level of the test.
+     */
+    nesting: number;
+    /**
+     * The ordinal number of the test.
+     */
+    testNumber: number;
+    /**
+     * Present if `context.todo` is called.
+     */
+    todo?: string | boolean;
+    /**
+     * Present if `context.skip` is called.
+     */
+    skip?: string | boolean;
+    /**
+     * The path of the test file, undefined if test is not ran through a file.
+     */
+    file?: string;
+}
+interface TestPass {
+    /**
+     * Additional execution metadata.
+     */
+    details: {
+        /**
+         * The duration of the test in milliseconds.
+         */
+        duration: number;
+    };
+    /**
+     * The test name.
+     */
+    name: string;
+    /**
+     * The nesting level of the test.
+     */
+    nesting: number;
+    /**
+     * The ordinal number of the test.
+     */
+    testNumber: number;
+    /**
+     * Present if `context.todo` is called.
+     */
+    todo?: string | boolean;
+    /**
+     * Present if `context.skip` is called.
+     */
+    skip?: string | boolean;
+    /**
+     * The path of the test file, undefined if test is not ran through a file.
+     */
+    file?: string;
+}
+interface TestPlan {
+    /**
+     * The nesting level of the test.
+     */
+    nesting: number;
+    /**
+     * The number of subtests that have ran.
+     */
+    count: number;
+    /**
+     * The path of the test file, undefined if test is not ran through a file.
+     */
+    file?: string;
+}
+interface TestStart {
+    /**
+     * The test name.
+     */
+    name: string;
+    /**
+     * The nesting level of the test.
+     */
+    nesting: number;
+    /**
+     * The path of the test file, undefined if test is not ran through a file.
+     */
+    file?: string;
+}
+interface TestStderr {
+    /**
+     * The path of the test file, undefined if test is not ran through a file.
+     */
+    file?: string;
+    /**
+     * The message written to `stderr`
+     */
+    message: string;
+}
+interface TestStdout {
+    /**
+     * The path of the test file, undefined if test is not ran through a file.
+     */
+    file?: string;
+    /**
+     * The message written to `stdout`
+     */
+    message: string;
+}
+
+/**
+ * The `node:test/reporters` module exposes the builtin-reporters for `node:test`.
+ * To access it:
+ *
+ * ```js
+ * import test from 'node:test/reporters';
+ * ```
+ *
+ * This module is only available under the `node:` scheme. The following will not
+ * work:
+ *
+ * ```js
+ * import test from 'test/reporters';
+ * ```
+ * @since v19.9.0
+ * @see [source](https://github.com/nodejs/node/blob/v20.2.0/lib/test/reporters.js)
+ */
+declare module 'node:test/reporters' {
+    import { Transform } from 'node:stream';
+
+    type TestEventName = 'diagnostic' | 'fail' | 'pass' | 'plan' | 'start' | 'stderr' | 'stdout';
+    type TestEventPayload = DiagnosticData | TestFail | TestPass | TestPlan | TestStart | TestStderr | TestStdout;
```

Comment:
How about making it more type-safe so that `switch/case` code can be used without using a type assertion: https://nodejs.org/dist/latest-v20.x/docs/api/test.html#custom-reporters ?

```ts
type TestEvent =
  | { type: 'test:diagnostic', data: DiagnosticData }
  | { type: 'test:fail', data: TestFail }
  ...

type TestEventGenerator = AsyncGenerator<TestEvent, void>;
```